internal/osbuild/assembler_test.go

```diff
@@ -0,0 +1,230 @@
+			},
+		},
+		{
+			name: "qemu assembler full",
+			assembler: Assembler{
+				Name: "org.osbuild.qemu",
+				Options: &QEMUAssemblerOptions{
+					Format:   "qcow2",
+					Filename: "disk.qcow2",
+					Size:     2147483648,
+					PTUUID:   "0x14fc63d2",
+					PTType:   "mbr",
+					Partitions: []QEMUPartition{QEMUPartition{
+						Start:    2048,
+						Bootable: true,
+						Filesystem: QEMUFilesystem{
+							Type:       "ext4",
+							UUID:       "76a22bf4-f153-4541-b6c7-0332c0dfaeac",
+							Label:      "root",
+							Mountpoint: "/",
+						},
+					}},
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.qemu","options":{"format":"qcow2","filename":"disk.qcow2","size":2147483648,"ptuuid":"0x14fc63d2","pttype":"mbr","partitions":[{"start":2048,"bootable":true,"filesystem":{"type":"ext4","uuid":"76a22bf4-f153-4541-b6c7-0332c0dfaeac","label":"root","mountpoint":"/"}}]}}`),
+			},
+		},
+		{
+			name: "tar assembler empty",
+			assembler: Assembler{
+				Name:    "org.osbuild.tar",
+				Options: &TarAssemblerOptions{},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.tar","options":{"filename":"","size":0}}`),
+			},
+		},
+		{
+			name: "tar assembler full",
+			assembler: Assembler{
+				Name: "org.osbuild.tar",
+				Options: &TarAssemblerOptions{
+					Filename:    "root.tar.xz",
+					Size:        0,
+					Compression: "xz",
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.tar","options":{"filename":"root.tar.xz","size":0,"compression":"xz"}}`),
+			},
+		},
+		{
+			name: "rawfs assembler empty",
+			assembler: Assembler{
+				Name:    "org.osbuild.rawfs",
+				Options: &RawFSAssemblerOptions{},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.rawfs","options":{"filename":"","root_fs_uuid":"00000000-0000-0000-0000-000000000000","size":0}}`),
+			},
+		},
+		{
+			name: "rawfs assembler full",
+			assembler: Assembler{
+				Name: "org.osbuild.rawfs",
+				Options: &RawFSAssemblerOptions{
+					Filename:           "filesystem.img",
+					RootFilesystemUUID: uuid.MustParse("76a22bf4-f153-4541-b6c7-0332c0dfaeac"),
+					Size:               2147483648,
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.rawfs","options":{"filename":"filesystem.img","root_fs_uuid":"76a22bf4-f153-4541-b6c7-0332c0dfaeac","size":2147483648}}`),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			assembler := &tt.assembler
+			var gotAssembler Assembler
+			err := gotAssembler.UnmarshalJSON(tt.args.data)
+			assert.Equalf((err != nil), tt.wantErr, "Assembler.UnmarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
+			if tt.wantErr {
+				return
+			}
+			gotBytes, err := json.Marshal(assembler)
+			assert.Nilf(err, "Could not marshal assembler: %v", err)
+			assert.Equalf(gotBytes, tt.args.data, "Expected `%v`, got `%v`", string(tt.args.data), string(gotBytes))
```

Comment:
The arguments order is expected, actual and the error message is really redundant. The assert framework should be doing this for you. If you swap the order of arguments then in case of assertion failure the message will be slightly different. 

You can experiment with this but IMO printing the 2 values as an error message is absolutely unnecessary. That's why we have the framework.

Reply:
That's a valid point, so I simplified all the `Equalf`s.